File: 3154-maximum-value-of-an-ordered-triplet-i/maximum-value-of-an-ordered-triplet-i.py
class Solution:
    def maximumTripletValue(self, nums: List[int]) -> int:
        n = len(nums)
        leftMax = [0] * n
        rightMax = [0] * n
        ans = 0

        for i in range(1, n):
            leftMax[i] = max(leftMax[i - 1], nums[i - 1])
            rightMax[n - 1 - i] = max(rightMax[n - i], nums[n - i])

        for j in range(1, n - 1):
            ans = max(ans, (leftMax[j] - nums[j]) * rightMax[j])

        return ans

File: 3154-maximum-value-of-an-ordered-triplet-i/maximum-value-of-an-ordered-triplet-i.cpp
class Solution {
public:
    long long maximumTripletValue(vector<int>& nums) {
        int n = nums.size();
        long long ans = 0;
        vector<int> leftMax(n), rightMax(n);

        for (int i = 1; i < n; i++) {
            leftMax[i] = max(leftMax[i - 1], nums[i - 1]);
            rightMax[n - 1 - i] = max(rightMax[n - i], nums[n - i]);
        }

        for (int j = 1; j < n - 1; j++) {
            ans = max(ans, (long long)(leftMax[j] - nums[j]) * rightMax[j]);
        }

        return ans;
    }
};